topics/eval-intro.md

```diff
@@ -0,0 +1,398 @@
+Scripts **should never** access keys with programmatically-generated names or based on the contents of data structures stored in the database.
+
+Any input to the function that isn't the name of a key is a regular input argument.
+
+In the example above, both _Hello_ and _Parameterization!_ regular input arguments for the script.
+Because the script doesn't touch any keys, we use the numerical argument _0_ to specify there are no key name arguments.
+The execution context makes arguments available to the script through [_KEYS_](/lua-api#the-keys-global-variable) and [_ARGV_](/lua-api#the-keys-global-variable) global runtime variables.
+The _KEYS_ table is pre-populated with all key name arguments provided to the script before its execution, whereas the _ARGV_ table serves a similar purpose but for regular arguments.
+
+The following attempts to demonstrate the distribution of input arguments between the scripts _KEYS_ and _ARGV_ runtime global variables:
+
+
+```
+redis> EVAL "return { KEYS[1], KEYS[2], ARGV[1], ARGV[2], ARGV[3] }" 2 key1 key2 arg1 arg2 arg3
+1) "key1"
+2) "key2"
+3) "arg1"
+4) "arg2"
+5) "arg3"
+```
+
+**Note:**
+as can been seen above, Lua's table arrays are returned as [RESP2 array replies](/topics/protocol#resp-arrays), so it is likely that your client's library will convert it to the native array data type in your programming language.
+Please refer to the rules that govern [ata type conversion](/topics/lua-api#data-type-conversion) for more pertinent information.
+
+## Interacting with Redis from a script
+
+It is possible to call Redis commands from a Lua script either via [`redis.call()`](/topics/lua-api#redis.call()) or [`redis.pcall()`](/topics/lua-api#redis.pcall()).
+
+The two are nearly identical.
+Both execute a Redis command along with its provided arguments, if these represent a well-formed command.
+However, the difference between the two functions lies in the manner in which runtime errors (such as syntax errors, for example) are handled.
+Errors raised from calling `redis.call()` function are returned directly to the client that had executed it.
+Conversely, errors encountered when calling the `redis.pcall()` function are returned to the script's execution context instead for possible handling.
+
+For example, consider the following:
+
+```
+> EVAL "return redis.call('SET', KEYS[1], ARGV[1])" 1 foo bar
+OK
+```
+The above script accepts one key name and one value as its input arguments.
+When executed, the script calls the `SET` command to set the input key, _foo_, with the string value "bar".
+
+## Script cache
+
+Until this point, we've used the `EVAL` command to run our scripts, with the emphasis that `EVAL` is a developer's command.
+
+Whenever we call `EVAL`, we also include the script's source code with the request.
+Repeatedly calling `EVAL` to execute the same set of parameterized scripts, wastes both network bandwidth and forces Redis to recompile the scripts every time.
```

Comment:
@oranagra if its the same script then its not recompile, but here the docs says `to execute the same set of parameterized scripts`. I understand it as if you put the parameters hard coded in the script, in such case it will be a new script each time and recompile.

Reply:
ohh, ok